File: problems/032/032.f90
PROGRAM F032

    ! CopyrIght 2021 MeLwyn FrancIs CarLo

    IMPLICIT NONE

    CHARACTER (LEN=10) :: TEMP_STR_1, TEMP_STR_2
    CHARACTER (LEN=20) :: TEMP_STR_3

    LOGICAL :: DUPLICATE_FOUND

    INTEGER, DIMENSION(1000) :: PRODUCTS_LIST = 0

    INTEGER :: SUM_VAL   = 0
    INTEGER :: COUNT_VAL = 1

    INTEGER :: I, J, K, L, PRODUCT_VAL, PRODUCT_LENGTH, TOTAL_LENGTH

    DO I = 1, 9999

        WRITE(TEMP_STR_1, '(I0)') I

        IF (INDEX(TEMP_STR_1, '0') /= 0) CYCLE

        DUPLICATE_FOUND = .FALSE.

        DO J = 1, LEN_TRIM(TEMP_STR_1)

            DO K = 1, LEN_TRIM(TEMP_STR_1)

                IF (J == K) CYCLE

                IF (TEMP_STR_1(J:J) == TEMP_STR_1(K:K)) THEN
                    DUPLICATE_FOUND = .TRUE.
                    EXIT
                END IF

            END DO

            IF (DUPLICATE_FOUND) EXIT

        END DO

        IF (DUPLICATE_FOUND) CYCLE

        DO J = 1, 9999

            WRITE(TEMP_STR_2, '(I0)') J

            IF (INDEX(TEMP_STR_2, '0') /= 0) CYCLE

            DUPLICATE_FOUND = .FALSE.

            DO K = 1, LEN_TRIM(TEMP_STR_2)

                DO L = 1, LEN_TRIM(TEMP_STR_2)

                    IF (K == L) CYCLE

                    IF (TEMP_STR_2(K:K) == TEMP_STR_2(L:L)) THEN
                        DUPLICATE_FOUND = .TRUE.
                        EXIT
                    END IF

                END DO

                IF (DUPLICATE_FOUND) EXIT

            END DO

            IF (DUPLICATE_FOUND) CYCLE

            DUPLICATE_FOUND = .FALSE.

            DO K = 1, LEN_TRIM(TEMP_STR_2)

                IF (INDEX(TEMP_STR_1, TEMP_STR_2(K:K)) /= 0) THEN
                    DUPLICATE_FOUND = .TRUE.
                    EXIT
                END IF

            END DO

            IF (DUPLICATE_FOUND) CYCLE

            PRODUCT_VAL = I * J
            PRODUCT_LENGTH = INT(LOG10(FLOAT(PRODUCT_VAL))) + 1

            WRITE(TEMP_STR_3, '(I0)') PRODUCT_VAL

            IF (INDEX(TEMP_STR_3, '0') /= 0) CYCLE

            TOTAL_LENGTH = PRODUCT_LENGTH                                      &
                         + LEN_TRIM(TEMP_STR_1)                                &
                         + LEN_TRIM(TEMP_STR_2)

            IF (TOTAL_LENGTH > 9) EXIT

            IF (TOTAL_LENGTH /= 9) CYCLE

            DUPLICATE_FOUND = .FALSE.

            DO K = 1, LEN_TRIM(TEMP_STR_3)

                DO L = 1, LEN_TRIM(TEMP_STR_3)

                    IF (K == L) CYCLE

                    IF (TEMP_STR_3(K:K) == TEMP_STR_3(L:L)) THEN
                        DUPLICATE_FOUND = .TRUE.
                        EXIT
                    END IF

                END DO

                IF (DUPLICATE_FOUND) EXIT

            END DO

            IF (DUPLICATE_FOUND) CYCLE

            DUPLICATE_FOUND = .FALSE.

            DO K = 1, PRODUCT_LENGTH

                IF (INDEX(TEMP_STR_1, TEMP_STR_3(K:K)) /= 0                    &
               .OR. INDEX(TEMP_STR_2, TEMP_STR_3(K:K)) /= 0) THEN

                    DUPLICATE_FOUND = .TRUE.
                    EXIT

                END IF

            END DO

            IF (DUPLICATE_FOUND) CYCLE

            PRODUCTS_LIST(COUNT_VAL) = PRODUCT_VAL

            COUNT_VAL = COUNT_VAL + 1

        END DO

    END DO

    DO I = 1, COUNT_VAL
        DO J = 1, COUNT_VAL
            IF (I == J) CYCLE
            IF (PRODUCTS_LIST(I) == PRODUCTS_LIST(J)) PRODUCTS_LIST(J) = 0
        END DO
    END DO

    DO I = 1, COUNT_VAL
        SUM_VAL = SUM_VAL + PRODUCTS_LIST(I)
    END DO

    PRINT ('(I0)'), SUM_VAL

END PROGRAM F032

File: problems/032/032.c
#include <math.h>
#include <stdio.h>
#include <string.h>

/* Copyright 2021 Melwyn Francis Carlo */

int main()
{
    int sum   = 0;
    int count = 0;

    int products_list[1000] = { 0 };

    for (int i = 1; i <= 9999; i++)
    {
        char temp_str_1[10] = { 0 };
        sprintf(&temp_str_1[0], "%d", i);

        if (strchr(temp_str_1, '0') != NULL)
            continue;

        int duplicate_found = 0;
        for (int j = 0; j < (int)strlen(temp_str_1); j++)
        {
            for (int k = 0; k < (int)strlen(temp_str_1); k++)
            {
                if (j == k)
                    continue;

                if (temp_str_1[j] == temp_str_1[k])
                {
                    duplicate_found = 1;
                    break;
                }
            }

            if (duplicate_found)
                break;
        }

        if (duplicate_found)
            continue;

        for (int j = 1; j <= 9999; j++)
        {
            char temp_str_2[10] = { 0 };
            sprintf(&temp_str_2[0], "%d", j);

            if (strchr(temp_str_2, '0') != NULL)
                continue;

            duplicate_found = 0;
            for (int k = 0; k < (int)strlen(temp_str_2); k++)
            {
                for (int l = 0; l < (int)strlen(temp_str_2); l++)
                {
                    if (k == l)
                        continue;

                    if (temp_str_2[k] == temp_str_2[l])
                    {
                        duplicate_found = 1;
                        break;
                    }
                }

                if (duplicate_found)
                    break;
            }

            if (duplicate_found)
                continue;

            duplicate_found = 0;
            for (int k = 0; k < (int)strlen(temp_str_2); k++)
            {
                if (strchr(temp_str_1, temp_str_2[k]) != NULL)
                {
                    duplicate_found = 1;
                    break;
                }
            }

            if (duplicate_found)
                continue;

            int product = i * j;
            int product_length = ((int)log10(product) + 1);
            char temp_str_3[product_length+1];
            memset(&temp_str_3[0], 0, product_length + 1);
            sprintf(&temp_str_3[0], "%d", product);

            if (strchr(temp_str_3, '0') != NULL)
                continue;

            int total_length = product_length 
                             + (int)strlen(temp_str_1) 
                             + (int)strlen(temp_str_2);

            if (total_length > 9)
                break;

            if (total_length != 9)
                continue;

            duplicate_found = 0;
            for (int k = 0; k < (int)strlen(temp_str_3); k++)
            {
                for (int l = 0; l < (int)strlen(temp_str_3); l++)
                {
                    if (k == l)
                        continue;

                    if (temp_str_3[k] == temp_str_3[l])
                    {
                        duplicate_found = 1;
                        break;
                    }
                }

                if (duplicate_found)
                    break;
            }

            if (duplicate_found)
                continue;

            int duplicate_found = 0;
            for (int k = 0; k < product_length; k++)
            {
                if ((strchr(temp_str_1, temp_str_3[k]) != NULL)
                ||  (strchr(temp_str_2, temp_str_3[k]) != NULL))
                {
                    duplicate_found = 1;
                    break;
                }
            }

            if (duplicate_found)
                continue;

            products_list[count++] = product;
        }
    }

    for (int i = 0; i < count; i++)
    {
        for (int j = 0; j < count; j++)
        {
            if (i == j)
                continue;

            if (products_list[i] == products_list[j])
                products_list[j] = 0;
        }
    }

    for (int i = 0; i < count; i++)
        sum += products_list[i];

    printf("%d\n", sum);

    return 0;
}
